packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts

```diff
@@ -329,48 +347,106 @@ export class Pipeline extends cdk.Construct implements cpapi.IPipeline {
     const pipelineRegion = this.node.stack.requireRegion(
-      "You need to specify an explicit region when using CodePipeline's cross-region support");
+        "You need to specify an explicit region when using CodePipeline's cross-region support");
 
     // if we already have an ArtifactStore generated for this region, or it's the Pipeline's region, nothing to do
-    if (this.artifactStores[action.region] || action.region === pipelineRegion) {
+    if (this.artifactStores[region] || region === pipelineRegion) {
       return;
     }
 
-    let replicationBucketName = this.crossRegionReplicationBuckets[action.region];
+    let replicationBucketName = this.crossRegionReplicationBuckets[region];
     if (!replicationBucketName) {
       const pipelineAccount = this.node.stack.requireAccountId(
-        "You need to specify an explicit account when using CodePipeline's cross-region support");
+          "You need to specify an explicit account when using CodePipeline's cross-region support");
       const app = this.node.stack.parentApp();
       if (!app) {
         throw new Error(`Pipeline stack which uses cross region actions must be part of an application`);
       }
       const crossRegionScaffoldStack = new CrossRegionScaffoldStack(app, {
-        region: action.region,
+        region,
         account: pipelineAccount,
       });
-      this._crossRegionScaffoldStacks[action.region] = crossRegionScaffoldStack;
+      this._crossRegionScaffoldStacks[region] = crossRegionScaffoldStack;
       replicationBucketName = crossRegionScaffoldStack.replicationBucketName;
     }
 
-    const replicationBucket = s3.Bucket.import(this, 'CrossRegionCodePipelineReplicationBucket-' + action.region, {
+    const replicationBucket = s3.Bucket.import(this, 'CrossRegionCodePipelineReplicationBucket-' + region, {
       bucketName: replicationBucketName,
     });
     replicationBucket.grantReadWrite(this.role);
 
-    this.artifactStores[action.region] = {
+    this.artifactStores[region] = {
       Location: replicationBucket.bucketName,
       Type: 'S3',
     };
   }
 
+  private attachActionForCrossAccount(action: cpapi.Action): iam.IRole | undefined {
+    if (!action.resource) {
+      return action.role;
+    }
+
+    const pipelineStack = this.node.stack;
+    const resourceStack = action.resource.node.stack;
+    if (pipelineStack.env.account !== resourceStack.env.account) {
+      if (!this.artifactBucket.encryptionKey) {
+        throw new Error('The Pipeline is being used in a cross-account manner, ' +
+            'but its artifact Bucket does not have a KMS Key defined. ' +
+            'A KMS key is required for a cross-account Pipeline. ' +
+            'Make sure to pass a Bucket with a Key when creating the Pipeline');
+      }
+
+      if (action.role) {
+        return action.role;
+      }
+
+      // make sure the Bucket has a physical name if the Construct itself is managing it
+      if (this.managedBucket) {
```

Comment:
Are you sure you want to do this *after* we've potentially early-returned because `action.role` was set?

Reply:
Changed this part, so I believe this is no longer relevant.